### PR TITLE
Include cleanup

### PR DESCRIPTION
**Pull request by MaxKellermann**
Two commits which acculumate to a 20% build speedup.

### CODE REVIEW

**Comment by MaxKellermann** [comment]
There were CI build failures because lots of includes were still missing. These were always missing, and hose bugs only now become visible with stricter includes. Will fix and force-push until all CI targets succeed the build.

**Comment by MaxKellermann** [comment]
Build time master / Fedora 30: 10m 04s
Build time this PR / Fedora 30: 5m 36s
(MacOS still doesn't build, but I'll do another force-push now.)

**Comment by MaxKellermann** [comment]
All CI tests succeed now (plus two false negatives). Phew.

**Comment by MaxKellermann** [comment]
Rebased and fixed merge conflict in `iosource/Manager.cc`.

File: src/Anon.cc [hunk 1]
```diff
@@ -1,14 +1,14 @@
...
-#include <sys/time.h>
-
+#include "Anon.h"
 #include "util.h"
 #include "net_util.h"
-#include "Anon.h"
 #include "Val.h"
 #include "NetVar.h"
 
+#include <stdlib.h>
```
Comment (timwoj) [inline src/Anon.cc]: The style guide specifies that system headers should come before all other headers (see https://github.com/zeek/zeek-docs/blob/master/devel/contributors/coding-style.rst#including-files). I'm inclined to adjust that to let the class header come before them. Is there another reason for moving the system headers to the end?

File: src/Anon.h [hunk 2]
```diff
@@ -11,11 +11,11 @@
 #pragma once
 
 #include <vector>
-#include <set>
 #include <map>
 
 #include "Reporter.h"
-#include "net_util.h"
+
+using std::map;
```
Comment (timwoj) [inline src/Anon.h]: We generally try to avoid `using` statements in headers, if possible, especially for `std` types. We had a PR recently that went through and removed a bunch of uses of `using namespace std` from all of the headers.

File: src/analyzer/protocol/tcp/TCP_Reassembler.h [hunk 3]
```diff
@@ -1,13 +1,18 @@
...
 #include "TCP_Endpoint.h"
 #include "TCP_Flags.h"
 
 class BroFile;
 class Connection;
 
-namespace analyzer { namespace tcp {
+namespace analyzer {
+
+class Analyzer;
```
Comment (timwoj) [inline src/analyzer/protocol/tcp/TCP_Reassembler.h]: Isn't this forward-declared in `analyzer/Analyzer.h` which you added?

File: src/logging/Tag.h [hunk 4]
```diff
@@ -3,13 +3,17 @@
...
 
 class EnumVal;
 
+namespace plugin {
+template <class T>
+class TaggedComponent;
+template <class T, class C>
+class ComponentManager;
+}
+
```
Comment (timwoj) [inline src/logging/Tag.h]: Redefining this block in multiple places (see also `input/Tag.h`) feels a little brittle, not that it's likely to ever change.

File: src/Anon.h (hunk 2 above)
Comment (MaxKellermann) [reply src/Anon.h]: Agree, I'd never do that - but where I added `using` statements, it was necessary to make the code build, because those sources depended on `using` statements in other headers which are no longer included.
Removing those `using` statements and adding namespace prefixes would be the next step.
But for this PR, I wanted to keep the code untouched as much as possible, and only change the includes.

File: src/Anon.cc (hunk 1 above)
Comment (MaxKellermann) [reply src/Anon.cc]: `Foo.cc` should always include `Foo.h` first to validate that `Foo.h` is self-containing and can be included by others without preconditions.
From there, it seems natural to me that "custom first, system last" is the natural order. We know/assert that system headers must be correct, so we include them in the order they need to be validated. But that's just a subjective opinion, and if it violates your coding style, I'll change it.

**Comment by jsiwek** [comment]
> * Do we want to standardize on where `zeek-config.h` gets included in the order of headers? In most places it's the first thing since it defines a bunch of configure-time options, but there's other places where that's not true.

Not something I'd probably remember to be too strict about, but yeah, my general expectation is that gets included "first, or as early as possible" such that it will already be in the right place if you need to use its definitions in subsequence platform-specific header inclusion logic.  So yes, worth suggesting that order as a standard practice.

> * Do we want to switch over to using the new "standard library" version of the C headers (cstdint, cmath, etc)?

I'd favor that, but not hugely significant IMO.  Also can be done independently/whenever.

File: src/Event.cc [hunk 5]
```diff
@@ -102,6 +103,26 @@ EventMgr::~EventMgr()
...
+	if ( h )
+		QueueEvent(new Event(h, std::move(vl), src, aid, mgr, obj));
+	else
+		{
+		for ( const auto& v : vl )
+			Unref(v);
+		}
+	}
+
+void EventMgr::Dispatch(Event* event, bool no_remote)
```
[the remark this answers is not in the document]
Comment (MaxKellermann) [reply src/Event.cc]: Will do.

**Comment by MaxKellermann** [comment]
>     * Do we want to standardize on where `zeek-config.h` gets included in the order of headers?

gcc and clang have the (non-standard) option `-include`, which the Linux kernel uses, for example. Zeek could decide to depend on that feature and thus force `zeek-config.h` to be always included in each compilation unit, without requiring this policy.
If you decide not to use that option, there should be a CI test which verifies that all (relevant) sources include this header first. Or else there may be interesting runtime breakages. This is dangerous and hard to debug.

>     * Do we want to switch over to using the new "standard library" version of the C headers (cstdint, cmath, etc)?

I don't have an opinion on this, but I think there's a `clang-tidy` option to do this automatically. If you decide to do so, I can give you a tree-wide PR for it.

File: src/Anon.cc (hunk 1 above)
Comment (timwoj) [reply src/Anon.cc]: The style is more along the lines of include-what-you-use, meaning if the source file needs it, it shouldn't be relying on some other header to include it. The easiest way to ensure that is the put them first. If some other header includes it, then include guards will prevent the double-include. Does that make enough sense?

I'll make a note to update the style document to put including the class header first.

File: src/Anon.h (hunk 2 above)
Comment (timwoj) [reply src/Anon.h]: Ok, that makes sense if you're going to follow-up and remove them later. 👍

**Comment by timwoj** [comment]
> gcc and clang have the (non-standard) option -include, which the Linux kernel uses, for example. Zeek could decide to depend on that feature and thus force zeek-config.h to be always included in each compilation unit, without requiring this policy.

That sounds totally fine to me.

> I'd favor that, but not hugely significant IMO. Also can be done independently/whenever.

I think the only concern is that it moves all of the methods into the std namespace, which will cause a lot of build breakage until they get sorted. It could lead to breakage for external plugins and whatnot too.

File: src/logging/Tag.h (hunk 4 above)
Comment (MaxKellermann) [reply src/logging/Tag.h]: Each header which needs a certain symbol should have a copy of the forward declaration, even if another header has already done it. Imagine that other header one day doesn't need the declaration anymore, or that header isn't included anymore - then `logging/Tag.h` must not break, and should not need to be adjusted. It should be self-contained.
Having multiple forward declarations for one symbol is allowed in C++.

**Comment by MaxKellermann** [comment]
> I think the only concern is that it moves all of the methods into the std namespace, which will cause a lot of build breakage until they get sorted.

True, Zeek would then need to refer to those symbols with `std::` qualification. Though in practice, the plain C symbols will be available as well in all C++ implementations I know, so I don't think anything will really (visibly) break. But in any case, for correctness, we should really use `std::`, but there's no immediate hurry.

> It could lead to breakage for external plugins and whatnot too.

If this breaks a plugin, then the plugin was already broken before, it just wasn't visible... having stricter semantics is a good thing, even if this reveals existing old breakages. Revealing mistakes means mistakes get fixed. That's good, even if it means some short-term annoyance.

File: src/Event.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (MaxKellermann) [reply src/Event.cc]: There's a problem with that: right now, the order in `Queue.cc` doesn't match the header already.
Header:
- `~EventMgr()`
- `Drain(...)`
- `QueueEvent(Event*)`
Source:
- `~EventMgr()`
- `QueueEvent(Event*)`
- `Drain()`
But anyway, I've moved it as you requested.

File: src/analyzer/protocol/tcp/TCP_Reassembler.h (hunk 3 above)
Comment (MaxKellermann) [reply src/analyzer/protocol/tcp/TCP_Reassembler.h]: Right, first I added the include to make it build, then I decided I can replace it with a forward declaration, but then forgot to remove the include... fixed now.

**Comment by MaxKellermann** [comment]
Force-pushed a new revision:
- rebased on master
- reverted the system include order according to the coding style documentation, with one exception: a file's own header is still included first
- removed the unnecessary new include from `TCP_Reassembler.h`

File: src/DNS_Mgr.h [hunk 6]
```diff
@@ -2,18 +2,16 @@
...
-#include "Dict.h"
 #include "EventHandler.h"
 #include "iosource/IOSource.h"
 #include "IPAddr.h"
 
+#include <list>
+#include <map>
+#include <queue>
+#include <utility>
+
```
Comment (timwoj) [inline src/DNS_Mgr.h]: This block needs to move up

File: src/util.cc [hunk 7]
```diff
@@ -42,6 +42,7 @@
 # include <malloc.h>
 #endif
 
+#include "Desc.h"
 #include "digest.h"
 #include "input.h"
 #include "util.h"
```
Comment (timwoj) [inline src/util.cc]: Move "util.h" up?

File: src/probabilistic/Hasher.cc [hunk 8]
```diff
@@ -1,13 +1,16 @@
...
-#include <openssl/evp.h>
-
 #include "Hasher.h"
 #include "NetVar.h"
 #include "digest.h"
 #include "siphash24.h"
 
+#include <broker/data.hh>
+
+#include <openssl/evp.h>
```
Comment (timwoj) [inline src/probabilistic/Hasher.cc]: Move this one and the one below it up in the order.

File: src/plugin/Plugin.h [hunk 9]
```diff
@@ -2,23 +2,21 @@
 
 #pragma once
 
+#include "zeek-config.h"
+#include "logging/WriterBackend.h"
```
Comment (timwoj) [inline src/plugin/Plugin.h]: Move this one down.

File: src/plugin/ComponentManager.h [hunk 10]
```diff
@@ -1,15 +1,15 @@
...
 #include "Type.h"
-#include "ID.h"
-#include "Var.h"
+#include "Var.h" // for add_type()
 #include "Val.h"
 #include "Reporter.h"
 #include "zeekygen/Manager.h"
+#include "DebugLogger.h"
+
+#include <map>
```
Comment (timwoj) [inline src/plugin/ComponentManager.h]: Move these up

File: src/iosource/PktSrc.h [hunk 11]
```diff
@@ -2,13 +2,16 @@
...
 #pragma once
 
-#include <vector>
-
 #include "IOSource.h"
-#include "BPF_Program.h"
-#include "Dict.h"
 #include "Packet.h"
 
+#include <vector>
```
Comment (timwoj) [inline src/iosource/PktSrc.h]: Move this one and the one below it up.

File: src/iosource/Manager.h [hunk 12]
```diff
@@ -4,12 +4,15 @@
 
 #include "zeek-config.h"
 
+#include "IOSource.h"
```
Comment (timwoj) [inline src/iosource/Manager.h]: Undo the change here to move these ones up.

File: src/file_analysis/File.h [hunk 13]
```diff
@@ -2,21 +2,22 @@
...
-
-#include "FileReassembler.h"
-#include "Conn.h"
-#include "Val.h"
-#include "Tag.h"
 #include "AnalyzerSet.h"
 #include "BroString.h"
 #include "WeirdState.h"
 
+#include <string>
```
Comment (timwoj) [inline src/file_analysis/File.h]: Move these up

File: src/broker/Manager.h [hunk 14]
```diff
@@ -1,5 +1,9 @@
 #pragma once
 
+#include "NetVar.h"
```
Comment (timwoj) [inline src/broker/Manager.h]: The internal headers and system headers need to be swapped in this one.

File: src/analyzer/protocol/icmp/ICMP.cc [hunk 15]
```diff
@@ -1,17 +1,20 @@
...
 #include "NetVar.h"
 #include "Event.h"
-#include "ICMP.h"
 #include "Conn.h"
+#include "Desc.h"
+#include "Reporter.h"
 
 #include "events.bif.h"
 
+#include <algorithm>
```
Comment (timwoj) [inline src/analyzer/protocol/icmp/ICMP.cc]: Move this up

File: src/analyzer/protocol/login/NVT.cc [hunk 16]
```diff
@@ -2,15 +2,16 @@
...
-
 #include "NVT.h"
 #include "NetVar.h"
 #include "Event.h"
+#include "Reporter.h"
 #include "analyzer/protocol/tcp/TCP.h"
 
 #include "events.bif.h"
 
+#include <stdlib.h>
```
Comment (timwoj) [inline src/analyzer/protocol/login/NVT.cc]: Move this up

File: src/analyzer/protocol/mime/MIME.h [hunk 17]
```diff
@@ -1,15 +1,18 @@
 #pragma once
 
+#include "BroString.h"
```
Comment (timwoj) [inline src/analyzer/protocol/mime/MIME.h]: Revert reordering  this one

File: src/analyzer/protocol/stepping-stone/SteppingStone.cc [hunk 18]
```diff
@@ -1,18 +1,17 @@
...
 #include "Net.h"
 #include "NetVar.h"
 #include "analyzer/protocol/tcp/TCP.h"
-#include "SteppingStone.h"
+#include "Sessions.h"
 #include "util.h"
-
 #include "events.bif.h"
 
+#include <stdlib.h>
```
Comment (timwoj) [inline src/analyzer/protocol/stepping-stone/SteppingStone.cc]: Move this up.

**Comment by MaxKellermann** [comment]
Rebased and fixed after @timwoj's comments
